src/equisolve/nn/module_tensor.py

```diff
@@ -0,0 +1,172 @@
+    from metatensor.core.labels import LabelsEntry
+
+    HAS_METATENSOR_TORCH = False
+
+from copy import deepcopy
+from typing import List, Optional
+
+import torch
+from torch.nn import Linear, Module, ModuleDict
+
+
+@torch.jit.interface
+class ModuleTensorMapInterface(torch.nn.Module):
+    """
+    This interface required for TorchScript to index the ModuleDict with non-literals
+    in ModuleTensorMap. Any module that is used with ModuleTensorMap must implement
+    this interface to be TorchScript compilable.
+
+    Note that the *typings and argument names must match exactly* so that an interface
+    is correctly implemented.
+
+    Reference
+    ---------
+    https://github.com/pytorch/pytorch/pull/45716
+    """
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        pass
+
+
+class ModuleTensorMap(Module):
+    """
+    ModuleDictTensorMap is needed when the keys correspond to different features sizes
+
+    :param module_map: a dictionary of modules with tensor map keys as dict keys
+        each module is applied on a block
+
+    :param out_tensor: a tensor map that has the output properties of the Labels
+    """
+
+    def __init__(self, module_map: ModuleDict, out_tensor: Optional[TensorMap] = None):
+        super().__init__()
+        self._module_map = module_map
+        self._out_tensor = out_tensor
+
+    @classmethod
+    def from_module(
+        cls,
+        in_keys: Labels,
+        module: Module,
+        many_to_one: bool = True,
+        out_tensor: Optional[TensorMap] = None,
+    ):
+        """
+        :param module: the mode that is applied on each block
+        :param many_to_one: specifies if a separate module for each block is used
+        :param out_tensor: a tensor map that has the output properties of the Labels
+                           and keys
+        """
+        module_map = ModuleDict()
+        for key in in_keys:
+            module_key = ModuleTensorMap.module_key(key)
+            if many_to_one:
+                module_map[module_key] = module
+            else:
+                module_map[module_key] = deepcopy(module)
+
+        return cls(module_map, out_tensor)
+
+    def forward(self, tensor: TensorMap) -> TensorMap:
+        out_blocks: List[TensorBlock] = []
+        for key, block in tensor.items():
+            out_block = self.forward_block(key, block)
+
+            for parameter, gradient in block.gradients():
+                if len(gradient.gradients_list()) != 0:
+                    raise NotImplementedError(
+                        "gradients of gradients are not supported"
+                    )
+                out_block.add_gradient(
+                    parameter=parameter,
+                    gradient=self.forward_block(key, gradient),
+                )
+            out_blocks.append(out_block)
+
+        return TensorMap(tensor.keys, out_blocks)
+
+    def forward_block(self, key: LabelsEntry, block: TensorBlock) -> TensorBlock:
```

Comment:
I think this can also be a Labels object right?

Reply:
if you get it by iterating over the keys, you'll get `LabelsEntry`